### DISCUSSION TITLE
How do I add debounce to `vim.schedule_wrap` or `vim.schedule`?

### DISCUSSION THREAD

**Discussion by nyngwang** [neovim/neovim#24038] [Q&A] [answered]
Hi, I'm trying to fix a plugin. This is my code: (In short, I'm trying to detect horizontal scrolling and call some `.update`. But I also want to avoid a burst of function calls)

```lua
vim.api.nvim_create_autocmd({ 'WinScrolled' }, {
    group = 'block.nvim',
    pattern = '*',
    callback = function(args)
        if vim.bo.buftype ~= '' then return end
        if not vim.v.event[args.match] then return end
        if vim.v.event.all.leftcol > 0 then
            vim.schedule(function () require("block").update(args.buf) end)
        end
    end
})
```

But if there is a burst of triggering of `if vim.v.event.all.leftcol > 0`, i.e. I want to detect scrolling {left,right}ward, then Neovim became very laggy.

Is it possible to be able to add some delay between my continuous `vim.schedule`-wrapped call like above so that `require('block').update(args.buf)` won't be called too many times within a short period? Or maybe we can abort some of them?

**Comment by zeertzjq** [comment, marked as answer]
Try using a timer:
```lua
local scroll_timer = nil

vim.api.nvim_create_autocmd({ 'WinScrolled' }, {
    group = 'block.nvim',
    pattern = '*',
    callback = function(args)
        if vim.bo.buftype ~= '' then return end
        if not vim.v.event[args.match] then return end
        if vim.v.event.all.leftcol > 0 then
          if scroll_timer then
            scroll_timer:close()
          end
          scroll_timer = vim.defer_fn(function()
            scroll_timer = nil
            require("block").update(args.buf)
          end, 500)
        end
    end
})
```

**Reply by haolian9** [reply]
maybe you guys already know this, but i feel i have to ask again and sorry for my ignorant: as `winscrolled` is triggered in a quite high fequency, is it an effecient way to keep creating & destroying timers?
